File: demo/dotnet/LeopardDemo/MicDemo.cs
/*
    Copyright 2020-2023 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
    file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
    specific language governing permissions and limitations under the License.
*/

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

using Pv;

namespace LeopardDemo
{
    /// <summary>
    /// Microphone Demo for Leopard Speech-to-Text engine.
    /// It creates an input audio stream from a microphone and processes it with Leopard. 
    /// </summary>
    public class MicDemo
    {
        private static readonly int PV_RECORDER_FRAME_LENGTH = 2048;

        /// <summary>
        /// Creates an input audio stream and instantiates an instance of Leopard object.
        /// </summary>
        /// <param name="accessKey">
        /// AccessKey obtained from Picovoice Console (https://console.picovoice.ai/).
        /// </param>
        /// <param name="modelPath">
        /// Absolute path to the file containing model parameters.
        /// If not set it will be set to the default location.
        /// </param>           
        /// <param name="enableAutomaticPunctuation">
        /// Set to `true` to enable automatic punctuation insertion.
        /// </param>
        /// <param name="enableSpeakerDiarization">
        /// Set to `true` to enable speaker diarization, which allows Leopard to differentiate speakers as
        /// part of the transcription process. Word metadata will include a `SpeakerTag` to identify unique speakers.
        /// </param>
        /// <param name="verbose">Enable verbose logging.</param>
        /// <param name="audioDeviceIndex">
        /// Optional argument. If provided, audio is recorded from this input device.
        /// Otherwise, the default audio input device is used.
        /// </param>        
        private static void RunDemo(
            string accessKey,
            string modelPath,
            bool enableAutomaticPunctuation,
            bool enableSpeakerDiarization,
            bool verbose,
            int audioDeviceIndex)
        {
            using (Leopard leopard = Leopard.Create(
                accessKey: accessKey,
                modelPath: modelPath,
                enableAutomaticPunctuation: enableAutomaticPunctuation,
                enableSpeakerDiarization: enableSpeakerDiarization))
            {
                using (PvRecorder recorder = PvRecorder.Create(PV_RECORDER_FRAME_LENGTH, audioDeviceIndex))
                {
                    Console.WriteLine($"Using device: {recorder.SelectedDevice}");
                    Console.CancelKeyPress += delegate (object sender, ConsoleCancelEventArgs e)
                    {
                        Console.WriteLine("Stopping...");
                    };

                    List<short> audioFrames = new List<short>();
                    recorder.Start();
                    Console.WriteLine(">>> Press `CTRL+C` to exit:\n");

                    while (true)
                    {
                        Console.WriteLine(">>> Recording ... Press `ENTER` to stop:");
                        var tokenSource = new CancellationTokenSource();
                        CancellationToken token = tokenSource.Token;
                        Task recordingTask = Task.Run(() =>
                        {
                            audioFrames.Clear();
                            recorder.Start();
                            while (!token.IsCancellationRequested)
                            {
                                short[] frame = recorder.Read();
                                audioFrames.AddRange(frame);
                            }
                            recorder.Stop();
                        });

                        string s = Console.ReadLine();
                        if (s == null)
                        {
                            break;
                        }

                        tokenSource.Cancel();
                        recordingTask.Wait();

                        short[] pcm = audioFrames.ToArray();

                        Console.WriteLine(">>> Processing ... \n");
                        try
                        {
                            LeopardTranscript result = leopard.Process(pcm);
                            Console.WriteLine(result.TranscriptString);
                            if (verbose)
                            {
                                Console.WriteLine(
                                    string.Format(
                                        "\n|{0,-15}|{1,-10:0.00}|{2,-10:0.00}|{3,-10:0.00}|{4,-10}|\n",
                                        "Word",
                                        "Confidence",
                                        "StartSec",
                                        "EndSec",
                                        "SpeakerTag"));
                                for (int i = 0; i < result.WordArray.Length; i++)
                                {
                                    LeopardWord word = result.WordArray[i];
                                    Console.WriteLine(
                                        string.Format(
                                            "|{0,-15}|{1,10:0.00}|{2,10:0.00}|{3,10:0.00}|{4,10}|",
                                            word.Word,
                                            word.Confidence,
                                            word.StartSec,
                                            word.EndSec,
                                            word.SpeakerTag));
                                }
                                Console.WriteLine();
                            }
                        }
                        catch (LeopardActivationLimitException)
                        {
                            Console.WriteLine($"AccessKey '{accessKey}' has reached its processing limit.");
                        }
                    }
                }
            }
        }

        /// <summary>
        /// Lists available audio input devices.
        /// </summary>
        private static void ShowAudioDevices()
        {
            string[] devices = PvRecorder.GetAvailableDevices();
            for (int i = 0; i < devices.Length; i++)
            {
                Console.WriteLine($"index: {i}, device name: {devices[i]}");
            }
        }

        public static void Main(string[] args)
        {
            AppDomain.CurrentDomain.UnhandledException += OnUnhandledException;
            if (args.Length == 0)
            {
                Console.WriteLine(HELP_STR);
                Console.Read();
                return;
            }

            string accessKey = null;
            string modelPath = null;
            bool enableAutomaticPunctuation = true;
            bool enableSpeakerDiarization = true;
            bool verbose = false;
            int audioDeviceIndex = -1;
            bool showAudioDevices = false;
            bool showHelp = false;

            // parse command line arguments
            int argIndex = 0;
            while (argIndex < args.Length)
            {
                if (args[argIndex] == "--access_key")
                {
                    if (++argIndex < args.Length)
                    {
                        accessKey = args[argIndex++];
                    }
                }
                else if (args[argIndex] == "--model_path")
                {
                    if (++argIndex < args.Length)
                    {
                        modelPath = args[argIndex++];
                    }
                }
                else if (args[argIndex] == "--disable_automatic_punctuation")
                {
                    enableAutomaticPunctuation = false;
                    argIndex++;
                }
                else if (args[argIndex] == "--disable_speaker_diarization")
                {
                    enableSpeakerDiarization = false;
                    argIndex++;
                }
                else if (args[argIndex] == "--verbose")
                {
                    verbose = true;
                    argIndex++;
                }
                else if (args[argIndex] == "--show_audio_devices")
                {
                    showAudioDevices = true;
                    argIndex++;
                }
                else if (args[argIndex] == "--audio_device_index")
                {
                    if (++argIndex < args.Length && int.TryParse(args[argIndex], out int deviceIndex))
                    {
                        audioDeviceIndex = deviceIndex;
                        argIndex++;
                    }
                }
                else if (args[argIndex] == "-h" || args[argIndex] == "--help")
                {
                    showHelp = true;
                    argIndex++;
                }
                else
                {
                    argIndex++;
                }
            }

            // print help text and exit
            if (showHelp)
            {
                Console.WriteLine(HELP_STR);
                Console.Read();
                return;
            }

            // print audio device info and exit
            if (showAudioDevices)
            {
                ShowAudioDevices();
                Console.Read();
                return;
            }

            // run demo with validated arguments
            RunDemo(
                accessKey,
                modelPath,
                enableAutomaticPunctuation,
                enableSpeakerDiarization,
                verbose,
                audioDeviceIndex);
        }

        private static void OnUnhandledException(object sender, UnhandledExceptionEventArgs e)
        {
            Console.WriteLine(e.ExceptionObject.ToString());
            Console.Read();
            Environment.Exit(1);
        }

        private static readonly string HELP_STR = "Available options: \n " +
            "\t--access_key (required): AccessKey obtained from Picovoice Console (https://console.picovoice.ai/)\n" +
            "\t--model_path: Absolute path to the file containing model parameters.\n" +
            "\t--audio_device_index: Index of input audio device.\n" +
            "\t--show_audio_devices: Print available recording devices.\n" +
            "\t--disable_automatic_punctuation: Disable automatic punctuation.\n" +
            "\t--disable_speaker_diarization: Disable speaker diarization.\n" +
            "\t--verbose: Enable verbose output. Prints Leopard word metadata.";
    }
}

File: demo/java/src/ai/picovoice/leoparddemo/MicDemo.java
/*
    Copyright 2022 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.leoparddemo;

import ai.picovoice.leopard.Leopard;
import ai.picovoice.leopard.LeopardTranscript;
import org.apache.commons.cli.*;

import javax.sound.sampled.*;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Recorder extends Thread {
    private TargetDataLine micDataLine = null;
    private boolean stop = false;
    private boolean isRecording = false;
    private List pcmBuffer = null;

    public Recorder(int audioDeviceIndex) {
        AudioFormat format = new AudioFormat(16000f, 16, 1, true, false);
        DataLine.Info dataLineInfo = new DataLine.Info(TargetDataLine.class, format);
        TargetDataLine micDataLine;
        try {
            micDataLine = getAudioDevice(audioDeviceIndex, dataLineInfo);
            micDataLine.open(format);
        } catch (LineUnavailableException e) {
            System.err.println("Failed to get a valid capture device. Use --show_audio_devices to " +
                    "show available capture devices and their indices");
            System.exit(1);
            return;
        }

        this.micDataLine = micDataLine;
        this.stop = false;
        this.isRecording = false;
        this.pcmBuffer = new ArrayList();
    }

    private static TargetDataLine getDefaultCaptureDevice(DataLine.Info dataLineInfo) throws LineUnavailableException {
        if (!AudioSystem.isLineSupported(dataLineInfo)) {
            throw new LineUnavailableException("Default capture device does not support the audio " +
                    "format required by Picovoice (16kHz, 16-bit, linearly-encoded, single-channel PCM).");
        }

        return (TargetDataLine) AudioSystem.getLine(dataLineInfo);
    }

    private static TargetDataLine getAudioDevice(
            int deviceIndex,
            DataLine.Info dataLineInfo) throws LineUnavailableException {
        if (deviceIndex >= 0) {
            try {
                Mixer.Info mixerInfo = AudioSystem.getMixerInfo()[deviceIndex];
                Mixer mixer = AudioSystem.getMixer(mixerInfo);

                if (mixer.isLineSupported(dataLineInfo)) {
                    return (TargetDataLine) mixer.getLine(dataLineInfo);
                } else {
                    System.err.printf(
                            "Audio capture device at index %s does not support the audio format required by " +
                                    "Picovoice. Using default capture device.", deviceIndex);
                }
            } catch (Exception e) {
                System.err.printf("No capture device found at index %s. Using default capture device.", deviceIndex);
            }
        }

        // use default capture device if we couldn't get the one requested
        return getDefaultCaptureDevice(dataLineInfo);
    }

    public void run() {
        this.isRecording = true;
        micDataLine.start();

        ByteBuffer captureBuffer = ByteBuffer.allocate(512);
        captureBuffer.order(ByteOrder.LITTLE_ENDIAN);
        short[] shortBuffer = new short[256];

        while (!stop) {
            micDataLine.read(captureBuffer.array(), 0, captureBuffer.capacity());
            captureBuffer.asShortBuffer().get(shortBuffer);
            for (int i = 0; i < shortBuffer.length; ++i) {
                this.pcmBuffer.add(shortBuffer[i]);
            }
        }
    }

    public void end() {
        this.stop = true;
    }

    public short[] getPCM() {
        short[] pcm = new short[this.pcmBuffer.size()];
        for (int i = 0; i < this.pcmBuffer.size(); ++i) {
            pcm[i] = (short) this.pcmBuffer.get(i);
        }
        return pcm;
    }
}

public class MicDemo {
    public static void runDemo(
            String accessKey,
            String modelPath,
            String libraryPath,
            boolean enableAutomaticPunctuation,
            boolean verbose,
            int audioDeviceIndex) {
        Leopard leopard = null;
        try {
            leopard = new Leopard.Builder()
                    .setAccessKey(accessKey)
                    .setModelPath(modelPath)
                    .setLibraryPath(libraryPath)
                    .setEnableAutomaticPunctuation(enableAutomaticPunctuation)
                    .build();

            System.out.println("Leopard version : " + leopard.getVersion());
            System.out.println(">>> Press `CTRL+C` to exit:");

            Recorder recorder = null;
            Scanner scanner = new Scanner(System.in);

            while (System.in.available() == 0) {
                if (recorder != null) {
                    System.out.println(">>> Recording ... Press 'ENTER' to stop:");
                    scanner.nextLine();
                    recorder.end();
                    while (recorder.isAlive()) {
                    }
                    short[] pcm = recorder.getPCM();

                    LeopardTranscript transcript = leopard.process(pcm);
                    System.out.println(transcript.getTranscriptString() + "\n");
                    if (verbose) {
                        LeopardTranscript.Word[] words = transcript.getWordArray();
                        System.out.format("%14s - %5s - %5s - %5s\n", "word", "start", "end", "confidence");
                        for (int i = 0; i < words.length; i++) {
                            System.out.format(
                                    "%2d: %10s - %5.2f - %5.2f - %5.2f\n",
                                    i,
                                    words[i].getWord(),
                                    words[i].getStartSec(),
                                    words[i].getEndSec(),
                                    words[i].getConfidence());
                        }
                    }
                    recorder = null;
                } else {
                    System.out.println(">>> Press 'ENTER' to start:");
                    scanner.nextLine();
                    recorder = new Recorder(audioDeviceIndex);
                    recorder.start();
                }
            }

            System.out.println("Stopping...");
        } catch (Exception e) {
            System.err.println(e.toString());
        } finally {
            if (leopard != null) {
                leopard.delete();
            }
        }
    }

    private static void showAudioDevices() {
        // get available audio devices
        Mixer.Info[] allMixerInfo = AudioSystem.getMixerInfo();
        Line.Info captureLine = new Line.Info(TargetDataLine.class);

        for (int i = 0; i < allMixerInfo.length; i++) {

            // check if supports capture in the format we need
            Mixer mixer = AudioSystem.getMixer(allMixerInfo[i]);
            if (mixer.isLineSupported(captureLine)) {
                System.out.printf("Device %d: %s\n", i, allMixerInfo[i].getName());
            }
        }
    }

    public static void main(String[] args) {
        Options options = buildCommandLineOptions();
        CommandLineParser parser = new DefaultParser();
        HelpFormatter formatter = new HelpFormatter();

        CommandLine cmd;
        try {
            cmd = parser.parse(options, args);
        } catch (ParseException e) {
            System.out.println(e.getMessage());
            formatter.printHelp("leopardmicdemo", options);
            System.exit(1);
            return;
        }

        if (cmd.hasOption("help")) {
            formatter.printHelp("leopardmicdemo", options);
            return;
        }

        if (cmd.hasOption("show_audio_devices")) {
            showAudioDevices();
            return;
        }

        String accessKey = cmd.getOptionValue("access_key");
        String modelPath = cmd.getOptionValue("model_path");
        String libraryPath = cmd.getOptionValue("library_path");
        boolean enableAutomaticPunctuation = !cmd.hasOption("disable_automatic_punctuation");
        boolean verbose = cmd.hasOption("verbose");
        String audioDeviceIndexStr = cmd.getOptionValue("audio_device_index");

        if (accessKey == null || accessKey.length() == 0) {
            throw new IllegalArgumentException("AccessKey is required for Leopard.");
        }

        if (libraryPath == null) {
            libraryPath = Leopard.LIBRARY_PATH;
        }

        if (modelPath == null) {
            modelPath = Leopard.MODEL_PATH;
        }

        int audioDeviceIndex = -1;
        if (audioDeviceIndexStr != null) {
            try {
                audioDeviceIndex = Integer.parseInt(audioDeviceIndexStr);
                if (audioDeviceIndex < 0) {
                    throw new IllegalArgumentException(String.format("Audio device index %s is not a " +
                            "valid positive integer.", audioDeviceIndexStr));
                }
            } catch (Exception e) {
                throw new IllegalArgumentException(String.format("Audio device index '%s' is not a " +
                        "valid positive integer.", audioDeviceIndexStr));
            }
        }

        runDemo(
                accessKey,
                modelPath,
                libraryPath,
                enableAutomaticPunctuation,
                verbose,
                audioDeviceIndex);
    }

    private static Options buildCommandLineOptions() {
        Options options = new Options();

        options.addOption(Option.builder("a")
                .longOpt("access_key")
                .hasArg(true)
                .desc("AccessKey obtained from Picovoice Console (https://console.picovoice.ai/).")
                .build());

        options.addOption(Option.builder("m")
                .longOpt("model_path")
                .hasArg(true)
                .desc("Absolute path to the file containing model parameters.")
                .build());

        options.addOption(Option.builder("l")
                .longOpt("library_path")
                .hasArg(true)
                .desc("Absolute path to the Leopard native runtime library.")
                .build());

        options.addOption(Option.builder("d")
                .longOpt("disable_automatic_punctuation")
                .desc("Disable automatic punctuation.")
                .build());

        options.addOption(Option.builder("di")
                .longOpt("audio_device_index")
                .hasArg(true)
                .desc("Index of input audio device.")
                .build());

        options.addOption(Option.builder("v")
                .longOpt("verbose")
                .desc("Enable verbose logging.")
                .build());

        options.addOption(new Option("sd", "show_audio_devices", false, "Print available recording devices."));
        options.addOption(new Option("h", "help", false, ""));

        return options;
    }
}
